Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp

```diff
@@ -113,178 +113,316 @@ static ThrowCompletionOr<Object*> array_species_create(VM& vm, Object& original_
     if (constructor.is_undefined())
         return TRY(Array::create(realm, length)).ptr();
 
+    // 7. If IsConstructor(C) is false, throw a TypeError exception.
     if (!constructor.is_constructor())
         return vm.throw_completion<TypeError>(ErrorType::NotAConstructor, TRY_OR_THROW_OOM(vm, constructor.to_string_without_side_effects()));
 
+    // 8. Return ? Construct(C, « 𝔽(length) »).
     return TRY(construct(vm, constructor.as_function(), Value(length))).ptr();
 }
 
 // 23.1.3.1 Array.prototype.at ( index ), https://tc39.es/ecma262/#sec-array.prototype.at
 JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::at)
 {
+    auto index = vm.argument(0);
+
+    // 1. Let O be ? ToObject(this value).
     auto this_object = TRY(vm.this_value().to_object(vm));
+
+    // 2. Let len be ? LengthOfArrayLike(O).
     auto length = TRY(length_of_array_like(vm, this_object));
-    auto relative_index = TRY(vm.argument(0).to_integer_or_infinity(vm));
-    if (Value(relative_index).is_infinity())
-        return js_undefined();
-    Checked<size_t> index { 0 };
+
+    // 3. Let relativeIndex be ? ToIntegerOrInfinity(index).
+    auto relative_index = TRY(index.to_integer_or_infinity(vm));
+
+    size_t k;
+
+    // 4. If relativeIndex ≥ 0, then
     if (relative_index >= 0) {
-        index += relative_index;
-    } else {
-        index += length;
-        index -= -relative_index;
+        // a. Let k be relativeIndex.
+        k = relative_index;
+    }
+    // 5. Else,
+    else {
+        // NOTE: Doing negative check for 6. here to avoid overflow.
+        if (length + relative_index < 0)
+            return js_undefined();
+
+        // a. Let k be len + relativeIndex.
+        k = length + relative_index;
     }
-    if (index.has_overflow() || index.value() >= length)
+
+    // 6. If k < 0 or k ≥ len, return undefined.
+    if (k >= length)
         return js_undefined();
-    return TRY(this_object->get(index.value()));
+
+    // 7. Return ? Get(O, ! ToString(𝔽(k))).
+    return TRY(this_object->get(k));
+}
+
+// 23.1.3.2.1 IsConcatSpreadable ( O ), https://tc39.es/ecma262/#sec-isconcatspreadable
+static ThrowCompletionOr<bool> is_concat_spreadable(VM& vm, Value const& val)
+{
+    // 1. If O is not an Object, return false.
+    if (!val.is_object())
+        return false;
+
+    // 2. Let spreadable be ? Get(O, @@isConcatSpreadable).
+    auto const& object = val.as_object();
+    auto spreadable = TRY(object.get(vm.well_known_symbol_is_concat_spreadable()));
+
+    // 3. If spreadable is not undefined, return ToBoolean(spreadable).
+    if (!spreadable.is_undefined())
+        return spreadable.to_boolean();
+
+    // 4. Return ? IsArray(O).
+    return TRY(val.is_array(vm));
 }
 
 // 23.1.3.2 Array.prototype.concat ( ...items ), https://tc39.es/ecma262/#sec-array.prototype.concat
 JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::concat)
 {
+    auto items = vm.running_execution_context().arguments;
```

Comment:
Hooked into arguments directly here and mutated it further down, because the spec said to `4. Prepend O to items.`.
Not sure if that's safe.

Reply:
Couldn't find a copy or clone method on `MarkedVector` unfortunately.